app/controllers/subscriptions_controller.rb

```diff
@@ -0,0 +1,35 @@
+class SubscriptionsController < ApplicationController
+  before_action :authenticate_user!
+
+  def index
+    @subscriptions = current_user.subscriptions
+  end
+
+  def create
+    # Expecting ransack query hash in params[:q]; store as JSON in query_params
+    qp = params[:q].is_a?(ActionController::Parameters) ? params[:q].to_unsafe_h : (params[:q].presence || {})
+    @subscription = current_user.subscriptions.build(query_params: qp)
+    if @subscription.save
+      if request.referer&.include?(listings_path)
+        redirect_to listings_path(q: qp), notice: "Arrival notification set."
+      else
+        redirect_to subscriptions_path, notice: "Subscription created successfully."
+      end
```

Comment:
You could here do a `redirect_back` to take the user back to where they were, and then fallback to the listing for the subscription.